# 1. Introduction.
Importing necessary libraries.    

In [2]:
import pandas as pd
import pandas_profiling as pp
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import describe


## 2. Loading data csv file into data_raw dataframe.

In [3]:
data_raw = pd.read_csv("car_prices.csv", error_bad_lines = False, low_memory = False, names= range(16))
type(data_raw)

/home/wildfire/.local/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)


pandas.core.frame.DataFrame

## 3. Checking head, tail and existing columns.

In [4]:
data_raw.head()

0      1         2           3      4             5                  6   \
0  year   make     model        trim   body  transmission                vin   
1  2015    Kia   Sorento          LX    SUV     automatic  5xyktca69fg566472   
2  2015    Kia   Sorento          LX    SUV     automatic  5xyktca69fg561319   
3  2014    BMW  3 Series  328i SULEV  Sedan     automatic  wba3c1c51ek116351   
4  2015  Volvo       S60          T5  Sedan     automatic  yv1612tb4f1310987   

      7          8         9      10        11  \
0  state  condition  odometer  color  interior   
1     ca          5     16639  white     black   
2     ca          5      9393  white     beige   
3     ca        4.5      1331   gray     black   
4     ca        4.1     14282  white     black   

                                       12     13            14  \
0                                  seller    mmr  sellingprice   
1                 kia motors america, inc  20500         21500   
2                 kia motors america, inc  20800         21500   
3  financial services remarketing (lease)  31900         30000   
4                 volvo na rep/world omni  27500         27750   

                                        15  
0                                 saledate  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
3  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
4  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)

In [13]:
eda = pp.ProfileReport(data_raw, title = "Used car auction prices exploration.")
eda

Summarize dataset:  74%|███████▍  | 23/31 [01:22<00:28,  3.59s/it, Missing diagram bar]


KeyboardInterrupt: 

In [5]:
data_raw.tail()

0       1       2            3          4          5   \
558833  2015     Kia    K900       Luxury      Sedan        NaN   
558834  2012     Ram    2500  Power Wagon   Crew Cab  automatic   
558835  2012     BMW      X5    xDrive35d        SUV  automatic   
558836  2015  Nissan  Altima        2.5 S      sedan  automatic   
558837  2014    Ford   F-150          XLT  SuperCrew  automatic   

                       6   7    8      9       10     11  \
558833  knalw4d4xf6019304  in  4.5  18255  silver  black   
558834  3c6td5et6cg112407  wa    5  54393   white  black   
558835  5uxzw0c58cl668465  ca  4.8  50561   black  black   
558836  1n4al3ap0fc216050  ga  3.8  16658   white  black   
558837  1ftfw1et2eke87277  ca  3.4  15008    gray   gray   

                                                       12     13     14  \
558833                                   avis corporation  35300  33000   
558834                                    i -5 uhlmann rv  30200  30800   
558835             financial services remarketing (lease)  29800  34000   
558836  enterprise vehicle exchange / tra / rental / t...  15100  11100   
558837                   ford motor credit company,llc pd  29600  26700   

                                             15  
558833  Thu Jul 09 2015 07:00:00 GMT-0700 (PDT)  
558834  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558835  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)  
558836  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)  
558837  Thu May 28 2015 05:30:00 GMT-0700 (PDT)

In [6]:
data_raw.shape

(558838, 16)

Our dataset has 558838 rows x 16 columns. Now we are going to make copy of the initial dataset.

In [7]:
used_car_prices_copy = data_raw.copy()
used_car_prices_copy = used_car_prices_copy.drop([used_car_prices_copy.index[0]])
used_car_prices_copy.head()

0      1                    2           3      4          5   \
1  2015    Kia              Sorento          LX    SUV  automatic   
2  2015    Kia              Sorento          LX    SUV  automatic   
3  2014    BMW             3 Series  328i SULEV  Sedan  automatic   
4  2015  Volvo                  S60          T5  Sedan  automatic   
5  2014    BMW  6 Series Gran Coupe        650i  Sedan  automatic   

                  6   7    8      9      10     11  \
1  5xyktca69fg566472  ca    5  16639  white  black   
2  5xyktca69fg561319  ca    5   9393  white  beige   
3  wba3c1c51ek116351  ca  4.5   1331   gray  black   
4  yv1612tb4f1310987  ca  4.1  14282  white  black   
5  wba6b2c57ed129731  ca  4.3   2641   gray  black   

                                       12     13     14  \
1                 kia motors america, inc  20500  21500   
2                 kia motors america, inc  20800  21500   
3  financial services remarketing (lease)  31900  30000   
4                 volvo na rep/world omni  27500  27750   
5  financial services remarketing (lease)  66000  67000   

                                        15  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
3  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
4  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
5  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)

In [8]:
used_car_prices_copy.columns=['Year', 'Brand', 'Model', 'Version', 'Body', 'Transmission', 'VIN', 'State', 'Condition rate', 
                   'Odometer', 'Color', 'Interior', 'Seller', 'MY', 'Price', 'Date']

In [9]:
used_car_prices_copy.head()

Year  Brand                Model     Version   Body Transmission  \
1  2015    Kia              Sorento          LX    SUV    automatic   
2  2015    Kia              Sorento          LX    SUV    automatic   
3  2014    BMW             3 Series  328i SULEV  Sedan    automatic   
4  2015  Volvo                  S60          T5  Sedan    automatic   
5  2014    BMW  6 Series Gran Coupe        650i  Sedan    automatic   

                 VIN State Condition rate Odometer  Color Interior  \
1  5xyktca69fg566472    ca              5    16639  white    black   
2  5xyktca69fg561319    ca              5     9393  white    beige   
3  wba3c1c51ek116351    ca            4.5     1331   gray    black   
4  yv1612tb4f1310987    ca            4.1    14282  white    black   
5  wba6b2c57ed129731    ca            4.3     2641   gray    black   

                                   Seller     MY  Price  \
1                 kia motors america, inc  20500  21500   
2                 kia motors america, inc  20800  21500   
3  financial services remarketing (lease)  31900  30000   
4                 volvo na rep/world omni  27500  27750   
5  financial services remarketing (lease)  66000  67000   

                                      Date  
1  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
2  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)  
3  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)  
4  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)  
5  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)

In [12]:
eda = pp.ProfileReport(used_car_prices_copy, title = "Used car auction prices exploration.")
eda

Summarize dataset:  52%|█████▏    | 23/44 [00:49<00:44,  2.14s/it, scatter df_index, df_index]


KeyboardInterrupt: 

Data frame shape: 558838 rows x 16 cols

#opis..

In [8]:
data_raw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 558811 entries, 0 to 558810
Data columns (total 16 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   year          558811 non-null  int64  
 1   make          548510 non-null  object 
 2   model         548412 non-null  object 
 3   trim          548160 non-null  object 
 4   body          545616 non-null  object 
 5   transmission  493458 non-null  object 
 6   vin           558811 non-null  object 
 7   state         558811 non-null  object 
 8   condition     547017 non-null  float64
 9   odometer      558717 non-null  float64
 10  color         558062 non-null  object 
 11  interior      558062 non-null  object 
 12  seller        558811 non-null  object 
 13  mmr           558811 non-null  int64  
 14  sellingprice  558811 non-null  int64  
 15  saledate      558811 non-null  object 
dtypes: float64(2), int64(3), object(11)
memory usage: 68.2+ MB


In [9]:
data_raw.describe()

year      condition       odometer            mmr  \
count  558811.000000  547017.000000  558717.000000  558811.000000   
mean     2010.038696       3.424512   68323.195797   13769.324646   
std         3.966812       0.949439   53397.752933    9679.874607   
min      1982.000000       1.000000       1.000000      25.000000   
25%      2007.000000       2.700000   28374.000000    7100.000000   
50%      2012.000000       3.600000   52256.000000   12250.000000   
75%      2013.000000       4.200000   99112.000000   18300.000000   
max      2015.000000       5.000000  999999.000000  182000.000000   

        sellingprice  
count  558811.000000  
mean    13611.262461  
std      9749.656919  
min         1.000000  
25%      6900.000000  
50%     12100.000000  
75%     18200.000000  
max    230000.000000

In [10]:
data_raw.isna().sum()

year                0
make            10301
model           10399
trim            10651
body            13195
transmission    65353
vin                 0
state               0
condition       11794
odometer           94
color             749
interior          749
seller              0
mmr                 0
sellingprice        0
saledate            0
dtype: int64

In [13]:
update(make_to_vin["make"])
df["vin_brand"]=df["vin"].apply(lambda x: x[0:3] )





In [26]:
make_to_vin=df[["make","vin_brand"]].drop_duplicates().dropna()
make_to_vin

make         0
vin_brand    0
dtype: int64

In [27]:
df.combine_first(make_to_vin["make"])
df

year    make                model         trim       body  \
0       2015     Kia              Sorento           LX        SUV   
1       2015     Kia              Sorento           LX        SUV   
2       2014     BMW             3 Series   328i SULEV      Sedan   
3       2015   Volvo                  S60           T5      Sedan   
4       2014     BMW  6 Series Gran Coupe         650i      Sedan   
...      ...     ...                  ...          ...        ...   
558806  2015     Kia                 K900       Luxury      Sedan   
558807  2012     Ram                 2500  Power Wagon   Crew Cab   
558808  2012     BMW                   X5    xDrive35d        SUV   
558809  2015  Nissan               Altima        2.5 S      sedan   
558810  2014    Ford                F-150          XLT  SuperCrew   

       transmission                vin state  condition  odometer   color  \
0         automatic  5xyktca69fg566472    ca        5.0   16639.0   white   
1         automatic  5xyktca69fg561319    ca        5.0    9393.0   white   
2         automatic  wba3c1c51ek116351    ca        4.5    1331.0    gray   
3         automatic  yv1612tb4f1310987    ca        4.1   14282.0   white   
4         automatic  wba6b2c57ed129731    ca        4.3    2641.0    gray   
...             ...                ...   ...        ...       ...     ...   
558806          NaN  knalw4d4xf6019304    in        4.5   18255.0  silver   
558807    automatic  3c6td5et6cg112407    wa        5.0   54393.0   white   
558808    automatic  5uxzw0c58cl668465    ca        4.8   50561.0   black   
558809    automatic  1n4al3ap0fc216050    ga        3.8   16658.0   white   
558810    automatic  1ftfw1et2eke87277    ca        3.4   15008.0    gray   

       interior                                             seller    mmr  \
0         black                            kia motors america, inc  20500   
1         beige                            kia motors america, inc  20800   
2         black             financial services remarketing (lease)  31900   
3         black                            volvo na rep/world omni  27500   
4         black             financial services remarketing (lease)  66000   
...         ...                                                ...    ...   
558806    black                                   avis corporation  35300   
558807    black                                    i -5 uhlmann rv  30200   
558808    black             financial services remarketing (lease)  29800   
558809    black  enterprise vehicle exchange / tra / rental / t...  15100   
558810     gray                   ford motor credit company,llc pd  29600   

        sellingprice                                 saledate vin_brand  
0              21500  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)       5xy  
1              21500  Tue Dec 16 2014 12:30:00 GMT-0800 (PST)       5xy  
2              30000  Thu Jan 15 2015 04:30:00 GMT-0800 (PST)       wba  
3              27750  Thu Jan 29 2015 04:30:00 GMT-0800 (PST)       yv1  
4              67000  Thu Dec 18 2014 12:30:00 GMT-0800 (PST)       wba  
...              ...                                      ...       ...  
558806         33000  Thu Jul 09 2015 07:00:00 GMT-0700 (PDT)       kna  
558807         30800  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)       3c6  
558808         34000  Wed Jul 08 2015 09:30:00 GMT-0700 (PDT)       5ux  
558809         11100  Thu Jul 09 2015 06:45:00 GMT-0700 (PDT)       1n4  
558810         26700  Thu May 28 2015 05:30:00 GMT-0700 (PDT)       1ft  

[558811 rows x 17 columns]